### PR TITLE
Add new example for section 5

### PR DESCRIPTION
**Pull request by idoleat**
A new directory /examples is added for example codes. .clang-format is copied from sysprog21/lkmpg project.

This draft PR provides a simplified implementation of thread pool. After initializing the thread pool with thread count, jobs can be added. The job queue is a SPMC ring buffer. To keep the implementation minimal, the producer is not protecting, resulting the thread pool can not run automatically when jobs are added. Or the worker may try to get the job before it is fully enqueued.

Padding is added in `thread_pool_t` to avoid false sharing. The number "40" is the sum of size of struct members, including alignment, before the first padding. There should be a better way to determine the value since structure packing is implementation defined.

The test in main function results non-determinate order of jobs echoing its id. A mechanism to wait all jobs to complete should be added later instead of using sleep. `thread_pool_destroyed()` is not functional yet.

To explain `Exchange`, `Test and set`, `Fetch and ...` and `Compare and swap` in section 5.1~5.4 using this example, the following issues should be resolved:
* `Exchange` is not in use.
* `Test and set` on the `initialized` flag is useless. Testing the flag on the first thread pool initialization retrieve non-determinate value. Currently re-initialization is still possible. If preventing re-initialization is not crucial in this example, we need to find another way to use test and set.
* Using `Fetch and add` to change thread pool state seems too intentional.
* Pick a better job type, such as BBP PI approximation.
* Finish implementation mentioned in the above paragraph.

Should we break the example into pieces to explain individually? Or list the code first?

### CODE REVIEW

File: examples/ch5_example.c [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+#include <stdatomic.h>
+#include <threads.h>
+#include <stdlib.h>
+#include <stdbool.h>
+
+#define CACHE_LINE_SIZE 64
+
+typedef struct job {
+    void *args;
+    struct job *next;
```
Comment (jserv) [inline examples/ch5_example.c]: Declare as `struct job *prev, *next;`

File: examples/ch5_example.c [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+int worker(void *args)
+{
+    if (!args)
+        return EXIT_FAILURE;
+    thread_pool_t *thrd_pool = (thread_pool_t *)args;
+
+    while (1) {
+        uintptr_t job;
+        if (atomic_load(&thrd_pool->state) == running) {
```
Comment (jserv) [inline examples/ch5_example.c]: Rewrite as following:
```c
if (atomic_load(&thrd_pool->state) == cancelled)
    return EXIT_SUCCESS;
if (atomic_load(&thrd_pool->state) == running) {
    ....
} else {
    continue;
}
```

File: examples/ch5_example.c [hunk 3]
```diff
@@ -36,32 +36,30 @@ int worker(void *args)
     thread_pool_t *thrd_pool = (thread_pool_t *)args;
 
     while (1) {
-        uintptr_t job;
-        if (atomic_load(&thrd_pool->state) == running) {
+        if (atomic_load(&thrd_pool->state) == cancelled) {
+            return EXIT_SUCCESS;
+        } else if (atomic_load(&thrd_pool->state) == running) {
```
Comment (jserv) [inline examples/ch5_example.c]: Here `else if` can shortened as a `if` statement.

File: examples/ch5_example.c [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+                // store happens-before while loop
+                atomic_store(&thrd_pool->state, idle);
+                while (1) {
+                    if (atomic_load(&thrd_pool->state) == running)
+                        break;
+                    // To auto run when jobs added, check if head and tail are different
+                    // as long as producer is protected
+                }
+            } else {
+                printf("Hello from job %d\n", *(int *)(*(job_t **)job)->args);
```
Comment (jserv) [inline examples/ch5_example.c]: Can you avoid using `*(int *)(*(job_t **)job)` by typedef?

File: examples/ch5_example.c [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        }
+    };
+}
+
+bool thread_pool_init(thread_pool_t *thrd_pool, int size)
+{
+    atomic_flag_test_and_set(&thrd_pool->initialezed); // It's useless anyway
+
+    // TODO: size should be a positive integer
+    // malloc with zero size is non-portable
```
Comment (jserv) [inline examples/ch5_example.c]: Simply place an `assert` to avoid the case size equals zero.

File: examples/ch5_example.c (hunk 4 above)
Comment (idoleat) [reply examples/ch5_example.c]: I think using macro could be better?
```c
#define CAST_JOB(job, type) ({ \
    *(typeof(type) *)(*(job_t **)(job))->args; \
})
```
so args in `printf` could be
```c
printf("Hello from job %d\n", CAST_JOB(job, int));
```

Comment (idoleat) [reply examples/ch5_example.c]: Also the above `(uintptr_t)(&(*(job_t **)job)->prev)` could be simplified to
```c
atomic_compare_exchange_strong(&thrd_pool->tail, &job, PREV_JOB(job))
```
by using
```c
#define PREV_JOB(job) ({ (uintptr_t)(&(*(job_t **)(job))->prev); })
```

**Comment by idoleat** [comment]
The job queue exhibits incorrect behavior and is currently under rework. To reproduce the incorrect behavior, add new jobs after finishing the current ones. A segmentation fault occurs simply because `thrd_pool->tail` is an independent variable, not a pointer to the real tail. It still holds the old address when new jobs are added. I overthought the situation, assuming that `prev` in `job_t` would be constantly changed, and thus made it a distinct struct member, even with padding.

**Comment by idoleat** [comment]
> the following issues should be resolved:
> * `Exchange` is not in use.
> * `Test and set` on the `initialized` flag is useless. Testing the flag on the first thread pool initialization retrieve non-determinate value. Currently re-initialization is still possible. If preventing re-initialization is not crucial in this example, we need to find another way to use test and set.
> * Using `Fetch and add` to change thread pool state seems too intentional.

Solved.

Operations on atomic types, including flag type, covered in C11 standard 7.17.7 and 7.17.8 are all used in the example code. Next I will start revising section 5 based on this example. The undesired result caused by not using atomic operations will be provided as well. Also clarifying more on why we need read-modify-write as a atomic step.

File: examples/ch5_example.c [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        thrd_pool->head->job.prev = &thrd_pool->head->job;
+    }
+
+    return true;
+}
+
+int main()
+{
+    thread_pool_t thrd_pool = { .initialezed = ATOMIC_FLAG_INIT };
+    int thread_count = 8;
```
Comment (jserv) [inline examples/ch5_example.c]: Both `job_count` and `thread_count` were meant to be constant, and you can simply specify them as macros. That is, `N_JOBS` and `N_THREADS`.

File: concurrency-primer.tex [hunk 7]
```diff
@@ -321,6 +321,9 @@ \section{Enforcing law and order}
 This model, defined by Leslie Lamport in 1979,
 is called \introduce{sequential consistency}.
 
+Notice that using atomic variables as an lvalue expression, such as \monobox{v\_ready = true} and \monobox{while(!v\_ready)}, is a convenient alternative to explicitly using \monobox{atomic\_load} or \monobox{atomic\_store}.
```
Comment (idoleat) [inline concurrency-primer.tex]: Code generation on atomic load and store is a little bit tricky. As explained in [LLVM's document](https://llvm.org/docs/Atomics.html#id16), sometimes they are just normal load/store, without atomic instructions, and their code generation can vary across different architectures. Should we explain more on this? Or readers may be confused on examples in section 7 and 10.1

File: concurrency-primer.tex [hunk 8]
```diff
@@ -321,6 +321,9 @@ \section{Enforcing law and order}
 This model, defined by Leslie Lamport in 1979,
 is called \introduce{sequential consistency}.
 
+Notice that using atomic variables as an lvalue expression, such as \monobox{v\_ready = true} and \monobox{while(!v\_ready)}, is a convenient alternative to explicitly using \monobox{atomic\_load} or \monobox{atomic\_store}. 
+Lvalue-to-rvalue conversion (which models a memory read from an atomic location to a CPU register) strips atomicity along with other qualifiers.
```
Comment (idoleat) [inline concurrency-primer.tex]: > This is from C11 6.7.2.4 and 6.7.3. Should I add reference as a footnote?

No, simply add in the text body.

File: concurrency-primer.tex [hunk 9]
```diff
@@ -443,52 +464,54 @@ \subsection{Compare and swap}
 It allows us to conditionally exchange a value \emph{if} its previous value matches some expected one.
 In \clang{} and \cplusplus{}, \textsc{CAS} resembles the following,
 if it were executed atomically:
-\begin{cppcode}
-template <typename T>
-bool atomic<T>::compare_exchange_strong(
-    T& expected, T desired)
+\begin{ccode}
```
Comment (idoleat) [inline concurrency-primer.tex:476]: This is borrowed from C11 7.17.4. Should I add reference as a footnote?

File: concurrency-primer.tex [hunk 10]
```diff
@@ -443,52 +464,54 @@ \subsection{Compare and swap}
...
-Let's say we have some long-running task that we might want to cancel.
-We'll give it three states: \textit{idle}, \textit{running},
-and \textit{cancelled}, and write a loop that exits when it is cancelled.
-\begin{cppcode}
-enum class TaskState : int8_t {
-    Idle, Running, Cancelled
-};
+Once threads are created in the thread pool as workers, they will continuously search for jobs to do.
+Jobs are taken from the tail of job queue.
+To claim a job without it being taken by another worker halfway through, we need to atomically change the pointer to the last job. Otherwise the last job is under races.
```
Comment (idoleat) [inline concurrency-primer.tex]: Should provide an example result of races with `-fsanitize=thread -g` enabled.

Comment (idoleat) [reply concurrency-primer.tex]: Thread sanitizer currently does not support C11 thread `[1]` `[2]` `[3]`*. Using functions in `threads.h` with `-fsanitize=thread` results in segmentation fault in thread sanitizer runtime. I will provide a minimum modification in git diff format in this subsection to run the example with Tsan using `pthread.h` instead.

`[1]`: https://clang.llvm.org/docs/ThreadSanitizer.html
`[2]`: https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual#faq
`[3]`: https://github.com/intel/libipt/commit/f24d71340f70dcdc2f738333bc6c09cec6509a81
*: Both documents [1] and [2] don't clearly specify that Tsan doesn't support C11 thread yet. By most-recent commit confirming this [3] and C++11 thread still requiring rebuilding libc++ with Tsan, I believe C11 is not supported in Tsan yet.

File: concurrency-primer.tex [hunk 11]
```diff
@@ -706,7 +753,17 @@ \section{Do we always need sequentially consistent operations?}
 they inhibit optimizations that your compiler and hardware would otherwise make.
 
 What if we could avoid some of this slowdown?
-Consider a simple case like the spinlock from \secref{spinlock}.
+Consider a simple case like spinlock.
+\begin{ccode}
```
Comment (idoleat) [inline concurrency-primer.tex]: `\secref{spinlock}` is originally one of the examples in section 5 and replaced by new example. Commit 2caa7c0 solved the missing reference by unrolling the example.

This is not a good example and is not simple. Spinning on RMW operation can cause cache bouncing. Test-and-test-and-set pattern is one way to solve this problem. Functionally, this is correct, but it may mislead readers by oversimplifying spin lock mechanics and ignoring the details behind it.

Spin lock topic should be moved to cache effect section. Here we simply use another example instead. Not sure if scheduling related topic, such as sched_yeild or spinning in user space, should be covered though. If so, a section dedicated to scheduling may be needed.

Comment (idoleat) [reply concurrency-primer.tex]: > Here we simply use another example instead

In 926f645,  spinlock example is replaced with new example provided in section 5.

File: concurrency-primer.tex [hunk 12]
```diff
@@ -853,51 +900,41 @@ \subsection{Relaxed}
...
 Although it may seem like a niche requirement, such scenarios are quite common.
 
-Refer back to our discussions on \secref{atomicity} and \secref{rmw} operations,
-where a worker thread increments a counter that a \textsc{UI} thread then reads.
-In this case, the counter can be incremented using \cc|fetch_add(1, memory_order_relaxed)| because the only requirement is atomicity;
-the counter itself does not coordinate synchronization.
-
-Relaxed operations are also beneficial for managing flags shared between threads.
-For example, a thread might continuously run until it receives a signal to exit:
+Relaxed operations are beneficial for managing flags shared between threads.
```
Comment (idoleat) [inline concurrency-primer.tex:948]: Should we mention the discussion around relaxed atomics?
like this one: https://lukegeeson.com/blog/2023-10-17-A-Proposal-For-Relaxed-Atomics/

**Comment by weihsinyeh** [comment]
To demonstrate the application of Test and Set in rmw_example.c:

As stated in the Concurrency Primer (§ 5.2) regarding Test and Set: "We could use Test and Set to build a simple spinlock."

In rmw_example.c, Compare and Swap is utilized to avoid race conditions. However, the Test and Set operation using atomic_flag_test_and_set() is only employed for initialization, which underutilizes its locking properties to avoid race condition.

To illustrate the distinct applications of Test and Set versus Compare and Swap:
Test and Set : If the Test operation fails, it will spin. 
Compare and Swap : When a Compare operation fails, it will continue to execute.

However, to discuss this in rmw_example.c, another shared resource must be introduced. Currently, the shared resource is a thrd_pool.

Consider a scenario where multiple workers all need to access the same shared variable to perform operations. In this case, mutex can be utilized to provide the functionality similar to a Test and Set flag.

In rmw_example.c, "Exchange, Test and Set, Fetch and ..., and Compare and Swap" performs all four operations using atomic operations from stdatomic.h. However, as mentioned in Chapter 1: "System programmers are familiar with tools such as mutexes, semaphores, and condition variables. Nevertheless, a question remains: How do these tools function, and how can we write concurrent code in their absence?"

Therefore, it is not mandatory to use atomic_flag_test_and_set() for Test and Set. Mutex can also achieve locking and unlocking effects.

**Comment by weihsinyeh** [comment]
Another idea for Chapter 5 is to retain the original structure, keeping the descriptions of concepts in sections 5.1, 5.2, 5.3, and 5.4. However, the original examples are not intuitive because all of these sections are actually related to RMW operations. It's important for readers to understand that they only differ in techniques, aimed at being applicable to different scenarios. At the same time, specifically explain where exactly the need for atomic operations arises.

Implementing RMW does not necessarily require tools like those found in stdatomic.h.

In a subsequent new subsection 5.5, further explanation will detail the atomic functions available to implement these four concepts. In rmw_example.c, demonstrate their application in specific scenarios.

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `main` branch for reviewing.

File: concurrency-primer.tex [hunk 13]
```diff
@@ -452,52 +473,78 @@ \subsection{Compare and swap}
...
+
+What if claiming a job, which updates \cc|thrd_pool->head->prev|, is not done atomically?
+Let's remove the atomicity of claiming a job and use thread sanitizer to detect races.
+Thread sanitizer is one of the sanitizers provided by \introduce{gcc} and \introduce{clang} to detect data races. 
+Data races are undefined behavior in \clang{}11 and \cplusplus{}11. 
+Thread sanitizer inserts runtime code into the target program to track memory accesses. 
+When races occur during execution, warning messages are printed.
+To enable this, add \monobox{-fsanitize=thread -g} to compiler flags.
+The following diff patch removes the atomicity of claiming a job and uses pthread instead of \clang{}11 thread, because thread sanitizer currently hasn't support \clang{}11 thread yet. 
+Save diff as \monobox{racer.diff} and patch the example code by \monobox{\$ patch rmw\_example.c race.diff}.
```
Comment (jserv) [inline concurrency-primer.tex]: Don't do that. This is a book, and we should emphasize the readability.

Comment (idoleat) [reply concurrency-primer.tex]: Got it. Would it be a better way to use conditional macros in the example code? like this:
```c
typedef struct tpool {
    atomic_flag initialezed;
    int size;
    atomic_int state;
#ifdef PTHREAD
    pthread_t *pool;
#else
    thrd_t *pool;
#endif
    idle_job_t *head;
} tpool_t;
```
then use a short a description guiding the reader to edit 3 lines of code to create race condition.

Comment (jserv) [reply concurrency-primer.tex]: Let's stick to C11 Threads rather than POSIX Thread.

File: concurrency-primer.tex (hunk 7 above)
Comment (jserv) [reply concurrency-primer.tex]: Can you summarize the above in the footnote?

File: concurrency-primer.tex [hunk 14]
```diff
@@ -380,39 +383,59 @@ \section{Arbitrarily-sized ``atomic'' types}
...
+So far we have introduced the importance of order and atomicity.
+In \secref{seqcst}, we see how an atomic object ensures the order of single store or load operation is not reordered by the compiler within a program. 
+Only upon establishing the correct inter-thread order can we continue to pursue how multiple threads can establish a correct cross-thread order. 
+After achieving this goal, we can further explore how concurrent threads can coordinate and collaborate smoothly.
+In \secref{atomicity}, there is a need for atomicity to ensure that a group of operations is not only sequentially executed but also completes without being interrupted by operation from other threads.
+This establishes correct order of operations from different threads.
+
+\includegraphics[keepaspectratio, width=0.6\linewidth]{images/atomic_rmw}
+\captionof{figure}{Exchange, Test and Set, Fetch and…, Compare and Swap can all be transformed into atomic RMW operations, ensuring that operations like t1 \to t2 \to t3 will become an atomic step.}
+\label{fig:atomic_rmw}
```
Comment (jserv) [inline concurrency-primer.tex]: Use `fig:atomic-rmw` for consistency.

File: concurrency-primer.tex [hunk 15]
```diff
@@ -431,73 +454,124 @@ \subsection{Test and set}
...
-  \item The worker thread loads the current counter value and adds one.
-  \item Before that thread can store the value back,
-        the \textsc{UI} thread zeroes the counter.
-  \item The worker now performs its store, as if the counter was never cleared.
-\end{enumerate}
+Transform \textsc{RMW} to directly modify the shared variable (such as addition, subtraction,
+or bitwise \textsc{AND}, \textsc{OR}, \textsc{XOR}) and return its previous value, 
+all as part of a single atomic operation. 
+Compare with \introduce{Exchange} \secref{exchange}, when programmers only need to make simple modification to the shared variable, 
+they can use \introduce{Fetch and…}.
```
Comment (jserv) [inline concurrency-primer.tex]: It is weird to have `Fetch and…`

File: concurrency-primer.tex (hunk 9 above)
Comment (jserv) [reply concurrency-primer.tex:476]: > This is borrowed from C11 7.17.4. Should I add reference as a footnote?

Add the proper statement.

File: concurrency-primer.tex (hunk 15 above)
Comment (weihsinyeh) [reply concurrency-primer.tex]: Could you please explain more?

File: concurrency-primer.tex (hunk 11 above)
Comment (idoleat) [reply concurrency-primer.tex]: Resolved as a6c68a9 restored the original content.

File: concurrency-primer.tex (hunk 7 above)
Comment (idoleat) [reply concurrency-primer.tex]: Resolved in 4dc5699.

File: concurrency-primer.tex (hunk 8 above)
Comment (idoleat) [reply concurrency-primer.tex]: Resolved in 4dc5699.

File: concurrency-primer.tex (hunk 9 above)
Comment (idoleat) [reply concurrency-primer.tex:476]: Resolved in 4dc5699.

File: concurrency-primer.tex (hunk 13 above)
Comment (idoleat) [reply concurrency-primer.tex]: Resolved in dcb8f9a.

File: examples/rmw_example.c [hunk 16]
```diff
@@ -0,0 +1,241 @@
...
+    idle_job->job.prev = &idle_job->job;
+    idle_job->prev = &idle_job->job;
+    thrd_pool->func = worker;
+    thrd_pool->head = idle_job;
+    thrd_pool->state = idle;
+    thrd_pool->size = size;
+
+    for (size_t i = 0; i < size; i++) {
+        thrd_create(thrd_pool->pool + i, worker, thrd_pool);
+        //TODO: error handling
```
Comment (jserv) [inline examples/rmw_example.c]: We may ignore some error handling for the sake of providing minimal examples.

File: concurrency-primer.tex (hunk 14 above)
Comment (idoleat) [reply concurrency-primer.tex]: Resolved in 11baa72.

**Comment by jserv** [comment]
Thank @idoleat for contributing!
